Add parallel horizon finder.

markscheel opened the pull request:
## Proposed changes

Adds horizon finder and an Integration Test for it.

The test finds two Schwarzschild horizons in parallel on two subsequent timesteps, on
a Cubed Sphere Domain with 48 Elements.

Depends on #755 and #794 (which now have been merged).

Depends on #960 and #962

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

markscheel commented:
I am hoping that generalizing the Horizon Components/Actions so that they (or some of them) can be also used for wave extraction can be done (by me) in a separate PR later, after this one is merged.  I am planning on doing wave extraction after horizons are in reasonable shape.

nilsvu commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,133 @@
+#include "Utilities/TMPL.hpp"
+
+/// Holds objects used by horizon finders.
+namespace Horizon {
```
Since this is a quite high-level namespace that's likely going to be used throughout the codebase we could go with an abbreviation such as `hrzn`. It's mostly up to you I guess @markscheel 😄

Also, I expect we'll add functionality to reconstruct event horizons from null geodesics at some point too, so we could name the directory `Horizons` instead of `ApparentHorizons`.

nilsvu commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,133 @@
+struct OptionHolder {
+  struct Verbosity {
+    using type = size_t;
+    static constexpr OptionString help = {"Verbosity"};
```
This should be part of a logger module that we haven't implemented yet (what's the status on that?).

kidder replied:
something planned, but no one is working on it...

nilsvu commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,133 @@
+/// for a single horizon, so that in the input file they can
+/// be specified under headings such as "AhA", "AhB", etc.
+template <typename Frame>
+struct OptionHolder {
```
Can this be in an anonymous namespace?

nilsvu commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,133 @@
+/// \details
+/// A HorizonComponent is responsible for finding a horizon and
+/// communicating with HorizonManagerComponents to do so.  There are
+/// only small number of HorizonComponents, one for each horizon
```
only _a_ small number of

nilsvu commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,133 @@
+///   - a type alias 'frame' indicating a frame from namespace ::Frame.
+///   - a type alias 'option_tag' containing something in Horizon::OptionTags.
+template <class Metavariables, typename AhTag>
+struct HorizonComponent {
```
Perhaps name the component `ApparentHorizonFinder`?

nilsvu commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+template <class Metavariables>
+struct HorizonManagerComponent;
+
+namespace Tags {
```
I think we decided to put `Tags` into their respective namespace, so e.g. `hrzn::Tags`.

kidder replied:
ah::Tags

nilsvu commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+/// Verbosity: 0 for silent, increasing values output increasing messages.
+struct Verbose : db::SimpleTag {
+  static std::string name() noexcept { return "Verbose"; }
+  using type = size_t;
```
There should be something like `Informer::Tags::Verbosity` eventually (see comment above). Do we want to keep logging functionality in this PR and replace it later?

kidder replied:
yes lets put the verbosity enum and tag in the Informer directory

nilsvu commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+  using type = ::FastFlow;
+};
+template <typename Frame>
+struct Strahlkorper : db::SimpleTag {
```
Perhaps `Shape`, since that's what the `Strahlkorper` describes?

nilsvu commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+
+    // Tell all the HorizonManagers about the convergence.
+    // Eventually when there are Actions for observers, control
+    // systems, etc. that depend on the horizon, they should be called
```
Are there multiple managers?

nilsvu replied:
Also, is the horizon finder responsible for broadcasting its results? Shouldn't it be the other way around, that dependent actions query and wait for the data?

kidder replied:
there is one manager per core.  And SpECTRE works by sending data, not querying in order to do task-based parallelism

nilsvu commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+}  // namespace Horizon
+
+namespace Actions {
+namespace Horizon {
```
Also swap, e.g.`hrzn::Actions`

nilsvu commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+namespace Horizon {
+
+template <typename AhTag>
+struct Initialize {
```
The other initializer actions follow the scheme e.g. `ApparentHorizonFinderTags::initialize` and also expose `simple_actions` and `compute_actions` type lists. I guess we should make a decision, document it and then be consistent with how we do databox initializations. I actually quite like the approach that you took here @markscheel because the initializer action can be used in `Parallel::simple_action`.

kidder replied:
@nilsleiffischer did you mean `simple_tags` and `compute_tags` ?  And, we don't understand your comment

kidder replied:
`simple_tags` and `compute_tags` should be exposed and `return_tag_list` should be a simple tmpl::append of the two

nilsvu commented on src/ApparentHorizons/HorizonManagerComponent.hpp:
```diff
@@ -0,0 +1,76 @@
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+
+namespace HorizonManager {
```
Perhaps just use the top-level horizon namespace? Or, if you want to put the tags below (`NumElements` and such) into their own namespace, how about `hrzn::Manager(::Tags)`?

kidder replied:
namespace ah

nilsvu commented on src/ApparentHorizons/HorizonManagerComponent.hpp:
```diff
@@ -0,0 +1,76 @@
+namespace HorizonManager {
+/// Holds all the options for a HorizonManager, so that
+/// they can be specified under a HorizonManager heading in the input file.
+struct OptionHolder {
```
Put in an anonymous namespace?

kidder replied:
no anonymous namespaces in header files, but can use  bla_detail

nilsvu commented on src/ApparentHorizons/HorizonManagerComponent.hpp:
```diff
@@ -0,0 +1,76 @@
+/// There should be one AhTag for each HorizonComponent.
+template <class Metavariables>
+struct HorizonManagerComponent {
+  using chare_type = Parallel::Algorithms::Group;
```
Out of curiosity, why is this a group chare?

markscheel replied:
I should document this better. The idea is that each Element sends its volume data only to the HorizonManagerComponent that lives on the same core as the Element, so that it's local communication.  Then after sending their data, the Elements go back to doing important things like computing right-hand sides for time steps, which is done concurrently with whatever horizon stuff is happening.  

The HorizonManagerComponents are responsible for interpolating data and communicating with the HorizonComponents.  Because horizon finding is iterative, there will typically be a handful of iterations for each horizon search, and each iteration requires communication back and forth between HorizonComponents and HorizonManagerComponents. This is done without any involvement of the Elements (except for the initial communication that needs to be done once for every horizon search, but not once for every iteration for every horizon search).

nilsvu replied:
Makes sense that elements just send their data and then keep computing other stuff. But could you use a singleton for collecting data and/or an array chare for parallelizing the horizon finding instead of the group chare? Or is it necessary to work on a particular core here? My current understanding is that we should leave the load balancing mostly up to Charm++.

kidder replied:
let's keep it as a group for now until we actually run something; a singleton would be overwhelmed by too much data.

nilsvu commented on src/ApparentHorizons/HorizonManagerComponent.hpp:
```diff
@@ -0,0 +1,76 @@
+///         - a type alias 'option_tag' to something in Horizon::OptionTags.
+/// There should be one AhTag for each HorizonComponent.
+template <class Metavariables>
+struct HorizonManagerComponent {
```
`ApparentHorizonManager`?

kidder replied:
ah::DataInterpolator ?  (and is more generic than just for horizons, so eventually shouldn't be in the ah namespace)

nilsvu commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,633 @@
+
+/// Simply prints the number of elements it talks to.
+/// Used for diagnostics.
+namespace HorizonManager {
```
swap namespaces

nilsvu commented on tests/Unit/ApparentHorizons/TestHorizonManagerHelper_ElementComponent.hpp:
```diff
@@ -0,0 +1,125 @@
+          *(global_cache.ckLocalBranch()));
+
+      Parallel::simple_action<Actions::DgElementArray::SendNumElements<
+          HorizonManagerComponent<Metavariables>>>(dg_element_array);
```
Can this be part of the initialization? For the linear solver I do a reduction across elements in the initialization phase so I think it's possible.

markscheel replied:
My current understanding is that the initialization function should not depend on having other parallel components already initialized, since each component is initialized in a random order.  In this case, SendNumElements calls an Action of HorizonManagerComponent, which needs to be initialized before it is called. If SendNumElements were called during the initialization, it might call an Action on an uninitialized HorizonManagerComponent.  Those that understand actions better than me should correct me if I am wrong.

nilsvu commented on tests/Unit/ApparentHorizons/TestHorizonManagerHelper_ElementComponent.hpp:
```diff
@@ -0,0 +1,125 @@
+          HorizonManagerComponent<Metavariables>>(
+          *(global_cache.ckLocalBranch()));
+      Parallel::simple_action<Actions::HorizonManager::PrintNumElements>(
+          my_proxy);
```
I suggest you add an action that does a few `SPECTRE_PARALLEL_REQUIRE`s to test results, like in `Test_AlgorithmReduction.cpp` for instance. In particular also in another phase after the horizon finding.

nilsvu reviewed (commented):
Cool 👍 This is a lot of code in one PR, so I started looking at architecture, namespaces and tests.

markscheel commented:
One reason why this PR is large is that it also includes  #755 and #794 (which it depends on).  As soon as  #755 and #794 are merged, this PR will be much smaller.

kidder commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,133 @@
+template <typename Frame>
+struct OptionHolder {
+  struct Verbosity {
+    using type = size_t;
```
make this an enum e.g. {Silent, Quiet, ..., BarfEverything} with a default of Silent/Quiet

kidder commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,133 @@
+  using initial_databox = db::compute_databox_type<
+      typename Actions::Horizon::Initialize<AhTag>::return_tag_list>;
+  using options =
+      tmpl::push_back<tmpl::list<typename AhTag::option_tag>,
```
tmpl::list<item1,item2>

kidder commented on src/ApparentHorizons/HorizonManagerComponent.hpp:
```diff
@@ -0,0 +1,76 @@
+/// they can be specified under a HorizonManager heading in the input file.
+struct OptionHolder {
+  struct Verbosity {
+    using type = size_t;
```
again an enum

kidder commented on src/ApparentHorizons/HorizonManagerComponent.hpp:
```diff
@@ -0,0 +1,76 @@
+  using options = tmpl::list<Verbosity>;
+  static constexpr OptionString help = {"Options for horizon manager"};
+
+  explicit OptionHolder(size_t&& verbosity_in) : verbosity(verbosity_in) {}
```
no && for simple types

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+
+// Returns the next set of points, in Frame coordinates, that data
+// should be interpolated onto.
+template <typename Frame>
```
put these helper functions in a detail namespace

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+  auto box = db::create<
+      db::AddSimpleTags<StrahlkorperTags::items_tags<Frame>>,
+      db::AddComputeTags<StrahlkorperTags::compute_items_tags<Frame>>>(
+      prolonged_strahlkorper);
```
std::move

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+
+  db::mutate<Tags::Horizon::IndicesOfFilledInterpPoints,
+             ::HorizonManager::VarsTags<frame>>(
+      make_not_null(&box), [&coords](auto indices_of_filled,
```
specify types for lambda args (for faster compile times)

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+  db::mutate<Tags::Horizon::FastFlow,
+             Tags::Horizon::Strahlkorper<typename AhTag::frame>>(
+      make_not_null(&box), [&inv_g, &ex_curv, &christ, &status_and_info ](
+                               auto fast_flow, auto strahlkorper) noexcept {
```
specify types

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+
+  const auto& status = status_and_info.first;
+  const auto& info = status_and_info.second;
+  if (verbose > 1 or (verbose > 0 && converged(status))) {
```
&& -> and

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+    // Prepare for finding horizon at a new time.
+    // For now, the initial guess for the new
+    // horizon is the result of the old one.
+    // Eventually HorizonChare will hold the previous horizons and
```
HorizonChare -> ah::Finder

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+    // Eventually HorizonChare will hold the previous horizons and
+    // will do time-extrapolation to set the next guess.
+    db::mutate<Tags::Horizon::FastFlow, Tags::Horizon::Timesteps>(
+        make_not_null(&box), [](auto fast_flow, auto steps) noexcept {
```
no auto

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+    ::Horizon::send_points_to_horizon_manager<AhTag, DbTags, Metavariables>(
+        box, cache, timesteps.front());
+  } else {
+    Parallel::printf("Finder %s failed, code = %s\n", AhTag::label(), status);
```
code -> reason

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+        db::get<Tags::Horizon::Timesteps>(box).empty();
+
+    db::mutate<Tags::Horizon::Timesteps>(
+        make_not_null(&box), [&timesteps](auto steps) noexcept {
```
no auto

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,372 @@
+    db::mutate<Tags::Horizon::IndicesOfFilledInterpPoints,
+               ::HorizonManager::VarsTags<frame>>(make_not_null(&box), [
+      &vars_src, &global_offsets
+    ](auto indices_of_filled, auto vars_dest) noexcept {
```
no auto

kidder reviewed (changes requested):
still need to review HorizonManagerComponentActions and the tests

group code review with @wthrowe @fmahebert @gsb76 @markscheel @mgiesler

markscheel committed:
Add Verbosity enum and tag.

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,633 @@
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
```
cond forward decls

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+  using type = size_t;
+};
+
+struct Psi : db::SimpleTag {
```
check gh to reuse these tags

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+  static std::string name() noexcept { return "ExCurv"; }
+};
+template <typename Frame>
+struct Christoffel2ndKind : db::SimpleTag {
```
PointwiseFunctions/GeneralRelativity/GrTags.hpp

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+}  // namespace Tags
+
+/// Variables that are sent from Elements to DataInterpolators.
+using VarsTagsFromElement =
```
tag alias are snake cased

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+/// on all elements associated with this DataInterpolator.
+struct VolumeVarsInfo : db::SimpleTag {
+  struct Info {
+    Element<3> element;
```
check if you can get away with only having element id

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+  };
+  // The size_t in the outer map is the timestep number.
+  using type =
+      std::unordered_map<size_t, std::unordered_map<ElementId<3>, Info>>;
```
recommendation is to use Time instead of size_t

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+  /// stored in this InterpolatedVarsInfo corresponds to a subset of
+  /// the points in block_coord_holders.  Moreover, the number of
+  /// interpolated points stored in this InterpolatedVarsInfo (the
+  /// size of 'vars' and 'global_offsets') will change as more
```
clarification on sizes described in parentheses

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+  /// InterpolatedVarsInfo lives only on a single core, and this core
+  /// will not have access to all the Elements).
+  std::vector<BlockLogicalCoordHolder<3>> block_coord_holders;
+  /// vars holds the variables on some subset of the points in
```
vars -> `vars`

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+  std::vector<BlockLogicalCoordHolder<3>> block_coord_holders;
+  /// vars holds the variables on some subset of the points in
+  /// block_coord_holders.  The grid points inside vars are indexed
+  /// linearly and arbitrarily.  The size of 'vars' changes as more
```
"...as determined by global_offsets"

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+
+/// Holds interpolated variables.
+template <typename Frame>
+struct InterpolatedVarsInfo {
```
place into dox group

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+  bool data_has_been_sent{false};
+};
+
+/// Holds interpolated variables at all times for a given AhType.
```
AhType -> AhTag

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+template <typename AhTag>
+struct HorizonHolder {
+  // The size_t in the unordered_map is the timestep number.
+  std::unordered_map<size_t, InterpolatedVarsInfo<typename AhTag::frame>>
```
size_t could go to double if we want to do time interpolation inside the horizon manager

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+struct make_tagged_tuple_impl;
+
+template <template <typename...> class List, typename... Tags>
+struct make_tagged_tuple_impl<List<Tags...>> {
```
See Nils' message

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+/// DataBox tag for a TaggedTuple containing all HorizonHolders.
+namespace Tags {
+template <typename Metavariables>
+struct horizon_holders : db::SimpleTag {
```
CamelCase

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+
+            intrp::Irregular<3> interpolator(
+                volume_info.mesh, element_coord_holder.element_logical_coords);
+            const auto interp_data = interpolator.interpolate(volume_info.vars);
```
Use vector of interp_data to avoid resizing variables every time this is called

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+
+// Sends interpolated data to a ah::Finder.
+template <typename Metavariables, typename DbTags, typename AhTag>
+void reduce_data(db::DataBox<DbTags>& box,
```
send_data_to_ah_finder

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+    auto& receiver_proxy =
+        Parallel::get_parallel_component<ah::Finder<Metavariables, AhTag>>(
+            cache);
+    Parallel::simple_action<Actions::Finder::ReceiveInterpolatedVars<AhTag>>(
```
check if clean up is done in the function calling this

markscheel replied:
Cleanup of volume data is done after all ah::Finders at that timestep have converged (so none of them need volume data).  Cleanup of interpolated data is done after a particular ah::Finder has converged, and when interpolated data at a new set of points is requested.  Some of the interpolated data could be cleaned up immediately after it is sent.

markscheel replied:
Now cleaning up earlier.

mar-celine commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,652 @@
+    const auto& phi = get<Tags::Phi>(vars);
+
+    // (what should I do instead?)
+    typename VarsTags<Frame::Inertial>::type output_vars(
```
db::item_type<VarsTags<Frame::Inertial>> output_vars(

markscheel committed:
Small fixes in GrTags, including adding SpatialChristoffelSecondKind.

kidder commented on src/ApparentHorizons/HorizonComponent.hpp:
```diff
@@ -0,0 +1,137 @@
+/// Holds objects used by horizon finders.
+namespace ah {
+
+// This struct groups together options that can be specified
```
move comment inside detail namespace

kidder commented on tests/Unit/ApparentHorizons/Test_HorizonManager.cpp:
```diff
@@ -0,0 +1,87 @@
+// IWYU pragma: no_forward_declare HorizonManagerComponent
+
+
+#include "Domain/DomainCreators/RegisterDerivedWithCharm.cpp"
```
include the hpp file as normal above

kidder commented on tests/Unit/ApparentHorizons/CMakeLists.txt:
```diff
@@ -20,3 +58,4 @@ add_test_library(
   "${LIBRARY_SOURCES}"
   "ApparentHorizons;EinsteinSolutions;RootFinding;Utilities"
   )
+
```
remove

kidder commented on tests/Unit/ApparentHorizons/TestHorizonManagerHelper_ElementActions.hpp:
```diff
@@ -0,0 +1,154 @@
+    auto& receiver_proxy =
+        Parallel::get_parallel_component<ParallelComponentOfReceiver>(cache);
+
+    ElementId<3> element_id{array_index};
```
const

kidder commented on tests/Unit/ApparentHorizons/TestHorizonManagerHelper_ElementComponent.hpp:
```diff
@@ -0,0 +1,127 @@
+      Parallel::CProxy_ConstGlobalCache<Metavariables>& global_cache,
+      std::unique_ptr<DomainCreator<3, Frame::Inertial>>
+          domain_creator) noexcept;
+  static void execute_next_phase(
```
define below

markscheel committed:
Add ah::DataInterpolator and ah::Finder and a test.

markscheel commented:
I just fixed some IWYU and clang-tidy problems (the latter because a clang-format moved // NOLINT to another line).

kidder reviewed (commented):
I forgot to push the button a long time ago, so ignore the ones you have already fixed  :)

kidder commented on src/Informer/Verbosity.hpp:
```diff
@@ -0,0 +1,32 @@
+
+#pragma once
+
+#include <ostream>
```
iosfwd

kidder commented on src/Informer/Verbosity.hpp:
```diff
@@ -0,0 +1,32 @@
+// Distributed under the MIT License.
```
Can you split off this commit into a separate PR, so it can be used in other PRs.  In doing that, can you also make a LoggingGroup, move the functions in Informer.hpp from UtilitiesGroup to LoggingGroup.

kidder commented on src/Informer/Verbosity.hpp:
```diff
@@ -0,0 +1,32 @@
+/// Indicates how much informative output a class should output.
+enum class Verbosity { Silent, Quiet, Verbose, Debug };
+
+namespace Tags {
```
should this be a CacheTag instead of a DataBox tag?

kidder replied:
Also move the tag to a separate file Tags.hpp

markscheel replied:
I put it in the DataBox inside the AH actions.  So doesn't it need to be a DataBox tag?

kidder replied:
Sorry, that was more of a design question.  In other words, isn't this a constant and therefore could it not be stored in the ConstGlobalCache instead?

markscheel replied:
If I do that, can I have different verbosities for AhA and AhC (e.g. make AhA quiet and AhC verbose)?  And a different verbosity for the Interpolator?

kidder commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,388 @@
+}
+}  // namespace Finder_detail
+
+namespace Actions {
```
Please split the individual actions into separate files.  Test each action individually with the action testing framework.
Then you can add each action as an individual commit and the actions can be a separate PR from the components.

kidder commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,597 @@
+}  // namespace Actions
+/// \endcond
+
+/// Tags describing items in a HorizonManager's DataBox.
```
Also put tags into a a Tags.hpp file

kidder commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,597 @@
+}
+}  // namespace DataInterpolator_detail
+
+namespace Actions {
```
again, split the actions into separate files, use the action testing framework,...

kidder reviewed (changes requested):
Group review with @nilsdeppe and myself

markscheel committed:
Add LoggingGroup and put Informer things into it.

markscheel commented:
I have rebased, and split off some of this into two new PRs (which are still included in this PR for now so that the tests pass, but will be removed later) but I haven't done anything with the action-testing framework yet.

moxcodes commented on src/ApparentHorizons/HorizonComponentActions.hpp:
```diff
@@ -0,0 +1,389 @@
+                                          db::item_type<Tags::Timesteps>*>
+                                              steps) noexcept {
+          steps->insert(steps->end(), timesteps.begin(), timesteps.end());
+        });
```
Maybe sort the steps in case they've been submitted in a weird order.

moxcodes reviewed (commented):
Group code review with @markscheel @mgiesler

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+namespace DataInterpolator_detail {
+// Interpolates data onto a set of points desired by a ah::Finder.
+template <typename Metavariables, typename DbTags, typename AhTag>
+void interpolate_data(db::DataBox<DbTags>& box, const Time& timestep) {
```
probably noexcept

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+
+          for (const auto& volume_info_inner : volume_info_outer.second) {
+            // Have we interpolated this element before?
+            if (interp_info.element_has_interpolated.find(
```
`unordered_set` count can make this simpler

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+  // Also, if we have already reduced data for this HorizonInfo, we should
+  // exit (we don't want to send the HorizonChare the same
+  // information twice).
+  if (interp_vars_info.find(timestep) == interp_vars_info.end() or
```
probably also can be a count

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+      make_not_null(&box),
+      [&timestep](const gsl::not_null<
+                  db::item_type<Tags::HorizonHolders<Metavariables>>*>
+                      holders_l) {
```
noexcept on lambda

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+
+    // Try to interpolate data for all AhTags
+    tmpl::for_each<typename Metavariables::horizon_tags>(
+        [&box, &cache, &timestep](auto x) {
```
noexcept

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+        [timestep, &block_coord_holders](
+            const gsl::not_null<
+                db::item_type<Tags::HorizonHolders<Metavariables>>*>
+                HorizonHolders) {
```
noexcept

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+        make_not_null(&box),
+        [&timestep](const gsl::not_null<
+                    db::item_type<Tags::HorizonHolders<Metavariables>>*>
+                        holders) {
```
noexcept

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+      using Tag = typename decltype(tag)::type;
+      const auto& found = get<HorizonHolderTag<Tag>>(holders)
+                              .time_steps_where_horizon_has_been_found;
+      if (found.find(timestep) == found.end()) {
```
count

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+      db::mutate<Tags::VolumeVarsInfo>(
+          make_not_null(&box),
+          [&timestep](const gsl::not_null<db::item_type<Tags::VolumeVarsInfo>*>
+                          volume_vars_info) {
```
noexcept

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+    // timestep, we will remove them.
+    bool this_timestep_is_done = true;
+    const auto& holders = db::get<Tags::HorizonHolders<Metavariables>>(box);
+    tmpl::for_each<typename Metavariables::horizon_tags>([&](auto tag) {
```
noexcept

moxcodes commented on src/ApparentHorizons/HorizonManagerComponentActions.hpp:
```diff
@@ -0,0 +1,598 @@
+        make_not_null(&box),
+        [&timestep](const gsl::not_null<
+                    db::item_type<Tags::HorizonHolders<Metavariables>>*>
+                        holders_l) {
```
noexcept

moxcodes reviewed (commented):
Group code review with @markscheel @mgiesler

kidder commented:
@markscheel can we close this now that #1208 has been merged?

nilsdeppe commented:
Closing this. Can be reopened if necessary.
